ocaml-lsp-server/src/document.ml

```diff
@@ -142,37 +145,43 @@ let make_config uri =
   Mconfig.get_external_config path mconfig
 
 let make_pipeline thread tdoc =
-  let async_make_pipeline =
-    Scheduler.async_exn thread (fun () ->
-        let text = Text_document.text tdoc in
-        let source = Msource.make text in
-        let config =
-          let uri = Text_document.documentUri tdoc in
-          make_config uri
+  ( lazy (Msource.make (Text_document.text tdoc))
+  , Lazy_fiber.create (fun () ->
+        let async_make_pipeline =
+          Scheduler.async_exn thread (fun () ->
+              let config =
+                let uri = Text_document.documentUri tdoc in
+                make_config uri
+              in
+              Mpipeline.make config (Msource.make (Text_document.text tdoc)))
         in
-        Mpipeline.make config source)
-  in
-  let open Fiber.O in
-  let+ res = await async_make_pipeline in
-  match res with
-  | Ok s -> s
-  | Error e -> Exn_with_backtrace.reraise e
+        let open Fiber.O in
+        let+ res = await async_make_pipeline in
+        match res with
+        | Ok s -> s
+        | Error e -> Exn_with_backtrace.reraise e) )
 
-let make timer merlin_thread tdoc =
+let make timer merlin_thread (tdoc : DidOpenTextDocumentParams.t) =
   let tdoc = Text_document.make tdoc in
-  (* we can do that b/c all text positions in LSP are line/col *)
-  let open Fiber.O in
-  let+ pipeline = make_pipeline merlin_thread tdoc in
-  { tdoc; pipeline; merlin = merlin_thread; timer }
+  let msource, pipeline = make_pipeline merlin_thread tdoc in
+  { tdoc; pipeline; merlin = merlin_thread; timer; msource }
 
-let update_text ?version doc changes =
-  let tdoc =
-    List.fold_left changes ~init:doc.tdoc ~f:(fun acc change ->
+let update_text ?version ({ merlin; tdoc; _ } as t) changes =
+  match
+    List.fold_left changes ~init:tdoc ~f:(fun acc change ->
         Text_document.apply_content_change ?version acc change)
-  in
-  let open Fiber.O in
-  let+ pipeline = make_pipeline doc.merlin tdoc in
-  { doc with tdoc; pipeline }
+  with
```

Comment:
This is interesting: I wonder whether we could save up on some computations sharing information across change applications. The k-th change is based on the state of document with (k-1)-th change applied, so it's not very obvious to me now.

Reply:
Probably the easiest thing to do that can speed this up is allocating the final string only once after `k` consecutive changes without a read. Right now we are creating a lot of GC pressure by throwing away a bunch of intermediate strings.